include/watchdog.h

```diff
@@ -1,116 +1,206 @@
-typedef int (*wdt_api_set_config)(struct device *dev,
-				  struct wdt_config *config);
-typedef void (*wdt_api_get_config)(struct device *dev,
-				   struct wdt_config *config);
-typedef void (*wdt_api_reload)(struct device *dev);
-
 struct wdt_driver_api {
-	wdt_api_enable enable;
-	wdt_api_disable disable;
-	wdt_api_get_config get_config;
-	wdt_api_set_config set_config;
-	wdt_api_reload reload;
+	int (*start)(struct device *dev);
+	int (*stop)(struct device *dev);
+	int (*configure)(struct device *dev, struct wdt_cfg *cfg);
+	int (*install_timeout)(struct device *dev, struct wdt_timeout_cfg *cfg);
+	int (*reload)(struct device *dev, void *handle);
 };
 
-static inline void wdt_enable(struct device *dev)
+/**
+ * @brief Enable watchdog instance.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 if successful.
+ * @retval -EFAULT if watchdog instance has not been configured yet.
+ * @retval -EBUSY if watchdog instance has been already started.
+ */
+static inline int wdt_start(struct device *dev)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->enable(dev);
+	return api->start(dev);
 }
 
-static inline void wdt_disable(struct device *dev)
+/**
+ * @brief Disable watchdog instance.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 if successful.
+ * @retval -EFAULT if watchdog instance is not enabled.
+ * @retval -EPERM if watchdog can not be stopped directly by application code.
+ */
+static inline int wdt_stop(struct device *dev)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->disable(dev);
+	return api->stop(dev);
 }
 
-static inline void wdt_get_config(struct device *dev,
-				  struct wdt_config *config)
+/**
+ * @brief Configure watchdog instance.
+ *	  This function is used for configuring global watchdog settings that
+ *	  affect all installed timeouts.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cfg Pointer to watchdog configuration structure.
+ *
+ * @retval 0 if successful.
+ * @retval -ENOTSUP if any of the set options settings is not supported.
+ */
+static inline int wdt_configure(struct device *dev, struct wdt_cfg *cfg)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->get_config(dev, config);
+	return api->configure(dev, cfg);
 }
 
-static inline int wdt_set_config(struct device *dev,
-				 struct wdt_config *config)
+/**
+ * @brief Install new timeout.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @pararm cfg Pointer to timeout configuration structure.
+ *
+ * @retval 0 if success.
+ * @retval -EBUSY if timeout can't be installed while watchdog has already
+ *		  been started.
+ * @retval -ENOMEM if no more timeouts can be installed.
+ * @retval -EINVAL if any of the window timeout value is out of possible range.
+ * @retval -ENOEXEC if WDT_MODE_CALLBACK is set but interrupt function
+ *		    pointer is NULL.
```

Comment:
I don't know it should be a terminal error here. Isn't it so the nrf5 wdt for instance can be set to trigger an interrupt on underflow/error (which could call a cb then) but will still reset anyway?

Maybe just a SYS_LOG_WARN() in this case would be just fine (to tell the dev that maybe he is doing something unwanted)